oxbow/src/bed.rs

```diff
@@ -0,0 +1,291 @@
+use arrow::{
+    array::{
+        ArrayRef, GenericStringBuilder, ListBuilder, StructBuilder, UInt16Builder, UInt64Builder,
+        UInt8Builder,
+    },
+    datatypes::{DataType, Field},
+    error::ArrowError,
+    record_batch::RecordBatch,
+};
+use noodles::bed;
+use std::{
+    fs::File,
+    io::{self, BufRead, BufReader},
+    str,
+    sync::Arc,
+};
+
+use crate::batch_builder::{write_ipc, BatchBuilder};
+
+pub fn new_from_path(path: &str) -> io::Result<bed::Reader<BufReader<File>>> {
+    File::open(path).map(BufReader::new).map(bed::Reader::new)
+}
+
+pub fn new_from_reader<R>(bed: R) -> io::Result<bed::Reader<BufReader<R>>>
+where
+    R: io::Read,
+{
+    Ok(bed::Reader::new(BufReader::new(bed)))
+}
+
+/// Returns the records in the given region as Apache Arrow IPC.
+///
+/// If the region is `None`, all records are returned.
+///
+/// # Examples
+///
+/// ```no_run
+/// use oxbow::bed;
+///
+/// let mut reader = bed::new_from_path("sample.bed").unwrap();
+/// let ipc = bed::records_to_ipc(reader).unwrap();
+/// ```
+pub fn records_to_ipc<T>(mut reader: bed::Reader<T>) -> Result<Vec<u8>, ArrowError>
+where
+    T: BufRead,
+{
+    let batch_builder = BedBatchBuilder::new(1024)?;
+    let records = reader.records().map(|r| r.unwrap());
+    write_ipc(records, batch_builder)
+}
+
+struct BedBatchBuilder {
+    reference_sequence_name: GenericStringBuilder<i32>,
+    start_position: UInt64Builder,
+    end_position: UInt64Builder,
+    name: GenericStringBuilder<i32>,
+    score: UInt16Builder,
+    strand: GenericStringBuilder<i32>,
+    thick_start: UInt64Builder,
+    thick_end: UInt64Builder,
+    color: StructBuilder,
+    block_sizes: ListBuilder<UInt64Builder>,
+    block_starts: ListBuilder<UInt64Builder>,
+}
+
+impl BedBatchBuilder {
+    pub fn new(_capacity: usize) -> Result<Self, ArrowError> {
+        Ok(Self {
+            reference_sequence_name: GenericStringBuilder::<i32>::new(),
+            start_position: UInt64Builder::new(),
+            end_position: UInt64Builder::new(),
+            name: GenericStringBuilder::<i32>::new(),
+            score: UInt16Builder::new(),
+            strand: GenericStringBuilder::<i32>::new(),
+            thick_start: UInt64Builder::new(),
+            thick_end: UInt64Builder::new(),
+            color: StructBuilder::new(
+                vec![
+                    Field::new("r", DataType::UInt8, true),
+                    Field::new("g", DataType::UInt8, true),
+                    Field::new("b", DataType::UInt8, true),
+                ],
+                vec![
+                    Box::new(UInt8Builder::new()),
+                    Box::new(UInt8Builder::new()),
+                    Box::new(UInt8Builder::new()),
+                ],
+            ),
```

Comment:
The `StructBuilder` for the arrow Struct Array takes the schema of the struct (the boolean sets the nullability of the field) and the builders for each field.